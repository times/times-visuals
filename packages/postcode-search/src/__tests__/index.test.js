import React from "react";
import { render, cleanup, fireEvent, wait } from "@testing-library/react";

import PostcodeSearch from "../index";

describe("<PostcodeSearch />", () => {
  beforeEach(() => {
    cleanup();
    fetch.resetMocks();
  });

  it("should render correctly", () => {
    const { asFragment } = render(<PostcodeSearch />);

    expect(asFragment()).toMatchSnapshot();
  });

  describe("postcodes.io", () => {
    it("should return the result from postcodes.io", async () => {
      fetch.mockResponses([
        JSON.stringify({
          result: {
            mock: "postcodes.io response"
          }
        }),
        { status: 200 }
      ]);

      const mockOnLoad = jest.fn();
      const { getByPlaceholderText, getByText } = render(
        <PostcodeSearch onLoad={mockOnLoad} />
      );

      fireEvent.change(getByPlaceholderText("Search for a postcode"), {
        currentTarget: {
          value: "SE19GF"
        }
      });
      fireEvent.click(getByText("Search"));

      await wait();

      expect(mockOnLoad).toHaveBeenCalledWith({
        mock: "postcodes.io response"
      });
    });

    it("should show an error if postcodes.io returns a non-200 response", async () => {
      fetch.mockResponses([
        JSON.stringify({
          error: "Foo bar"
        }),
        { status: 500 }
      ]);

      const mockOnError = jest.fn();
      const { getByPlaceholderText, getByText } = render(
        <PostcodeSearch onError={mockOnError} />
      );

      fireEvent.change(getByPlaceholderText("Search for a postcode"), {
        currentTarget: {
          value: "SE19GF"
        }
      });
      fireEvent.click(getByText("Search"));

      await wait();

      expect(mockOnError).toHaveBeenCalledWith(
        new Error("Internal Server Error")
      );
    });
  });

  describe("custom provider", () => {
    it("should return the result from the custom provider function", async () => {
      const mockOnLoad = jest.fn();
      const { getByPlaceholderText, getByText } = render(
        <PostcodeSearch
          provider="custom"
          onLoad={mockOnLoad}
          onRequest={() =>
            Promise.resolve({
              mock: "custom provider response"
            })
          }
        />
      );

      fireEvent.change(getByPlaceholderText("Search for a postcode"), {
        currentTarget: {
          value: "SE19GF"
        }
      });
      fireEvent.click(getByText("Search"));

      await wait();

      expect(mockOnLoad).toHaveBeenCalledWith({
        mock: "custom provider response"
      });
    });

    it("should show an error if the custom provider function rejects", async () => {
      const mockOnError = jest.fn();
      const { getByPlaceholderText, getByText } = render(
        <PostcodeSearch
          provider="custom"
          onError={mockOnError}
          onRequest={() => Promise.reject(new Error("Some error"))}
        />
      );

      fireEvent.change(getByPlaceholderText("Search for a postcode"), {
        currentTarget: {
          value: "SE19GF"
        }
      });
      fireEvent.click(getByText("Search"));

      await wait();

      expect(mockOnError).toHaveBeenCalledWith(new Error("Some error"));
    });

    it("should show an error if no custom provider request function is provided", async () => {
      const mockOnError = jest.fn();
      const { getByPlaceholderText, getByText } = render(
        <PostcodeSearch provider="custom" onError={mockOnError} />
      );

      fireEvent.change(getByPlaceholderText("Search for a postcode"), {
        currentTarget: {
          value: "SE19GF"
        }
      });
      fireEvent.click(getByText("Search"));

      await wait();

      expect(mockOnError).toHaveBeenCalledWith(
        new Error(
          "Please provide the `onRequest` prop when using a custom provider"
        )
      );
    });
  });
});
