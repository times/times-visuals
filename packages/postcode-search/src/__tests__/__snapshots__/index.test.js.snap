// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PostcodeSearch /> should render correctly 1`] = `
<DocumentFragment>
  .c0 form {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c0 input[type="search"] {
  width: 100%;
  margin-top: 0;
  margin-bottom: 0;
  border: 1px solid #dbdbdb;
  font-size: 1.6rem;
  line-height: 1.8rem;
  padding: 0.8rem 0.9rem 0.8rem;
  text-transform: uppercase;
}

.c0 input[type="search"]::-webkit-input-placeholder {
  text-transform: none;
}

.c0 input[type="search"]::-moz-placeholder {
  text-transform: none;
}

.c0 input[type="search"]:-ms-input-placeholder {
  text-transform: none;
}

.c0 input[type="search"]::placeholder {
  text-transform: none;
}

.c0 input[type="submit"] {
  margin: 0;
  min-width: 12rem;
  padding: 1.5rem 0.9rem 1rem;
  color: #fff !important;
  font-size: 1.2rem;
  font-family: "GillSansMTStd-Medium","GillSans-Medium";
  text-transform: uppercase;
  border-radius: 0;
  background: #046798;
  box-shadow: none;
  -webkit-appearance: none;
}

.c0 input[type="submit"]:hover,
.c0 input[type="submit"]:focus {
  background: darken(#046798,15%);
  border: 0;
  box-shadow: none;
}

.c0 input[type="submit"]:disabled {
  opacity: 0.7;
}

@media screen and (min-width:768px) {
  .c0 input[type="search"] {
    margin-top: 0;
    margin-bottom: 0;
  }
}

@media screen and (min-width:768px) {
  .c0 input[type="submit"] {
    margin: 1rem 0;
    font-size: 1.6rem;
    margin-top: 0;
    margin-bottom: 0;
  }
}

<div
    class="c0"
  >
    <form>
      <input
        autocomplete="off"
        name="location"
        placeholder="Search for a postcode"
        type="search"
        value=""
      />
      <input
        class="Button"
        type="submit"
        value="Search"
      />
    </form>
  </div>
</DocumentFragment>
`;
